swap/peer.go

```diff
@@ -0,0 +1,223 @@
+			return
+		}
+		log.Info("submit tx minded", "receipt", receipt)
+
+		receipt, err = otherSwap.CashChequeBeneficiary(opts, sp.backend, sp.swap.owner.Contract, big.NewInt(int64(actualAmount)))
+		if err != nil {
+			log.Error("Got error when calling cashChequeBeneficiary", "err", err)
+			//TODO: do something with the error
+			return
+		}
+		log.Info("cash tx minded", "receipt", receipt)
+		//TODO: after the cashCheque is done, we have to watch the blockchain for x amount (25) blocks for reorgs
+		//TODO: make sure we make a case where we listen to the possibiliyt of the peer shutting down.
+	}()
+	return err
+}
+
+// TODO: Error handling
+// handleErrorMsg is called when an ErrorMsg is received
+func (sp *Peer) handleErrorMsg(ctx context.Context, msg interface{}) error {
+	log.Info("received error msg")
+	// maybe balance disagreement
+	return nil
+}
+
+// processAndVerifyCheque verifies the cheque and compares it with the last received cheque
+// if the cheque is valid it will also be saved as the new last cheque
+func (sp *Peer) processAndVerifyCheque(cheque *Cheque) (uint64, error) {
+	if err := sp.verifyChequeProperties(cheque); err != nil {
+		return 0, err
+	}
+
+	lastCheque := sp.loadLastReceivedCheque()
+
+	// TODO: there should probably be a lock here?
+	expectedAmount, err := sp.swap.oracle.GetPrice(cheque.Honey)
+	if err != nil {
+		return 0, err
+	}
+
+	actualAmount, err := sp.verifyChequeAgainstLast(cheque, lastCheque, expectedAmount)
+	if err != nil {
+		return 0, err
+	}
+
+	if err := sp.saveLastReceivedCheque(cheque); err != nil {
+		log.Error("error while saving last received cheque", "peer", sp.ID().String(), "err", err.Error())
+		// TODO: what do we do here?
+	}
+
+	return actualAmount, nil
+}
+
+// verifyChequeProperties verifies the signautre and if the cheque fields are appropiate for this peer
+// it does not verify anything that requires knowing the previous cheque
+func (sp *Peer) verifyChequeProperties(cheque *Cheque) error {
+	if cheque.Contract != sp.contractAddress {
+		return fmt.Errorf("wrong cheque parameters: expected contract: %x, was: %x", sp.contractAddress, cheque.Contract)
+	}
+
+	// the beneficiary is the owner of the counterparty swap contract
+	if err := cheque.VerifySig(sp.beneficiary); err != nil {
+		return err
+	}
+
+	if cheque.Beneficiary != sp.swap.owner.address {
+		return fmt.Errorf("wrong cheque parameters: expected beneficiary: %x, was: %x", sp.swap.owner.address, cheque.Beneficiary)
+	}
+
+	if cheque.Timeout != 0 {
+		return fmt.Errorf("wrong cheque parameters: expected timeout to be 0, was: %d", cheque.Timeout)
+	}
+
+	return nil
+}
+
+// verifyChequeAgainstLast verifies that serial and amount are higher than in the previous cheque
+// furthermore it cheques that the increase in amount is as expected
+// returns the actual amount received in this cheque
+func (sp *Peer) verifyChequeAgainstLast(cheque *Cheque, lastCheque *Cheque, expectedAmount uint64) (uint64, error) {
```

Comment:
Additionally, shouldn't the lastCheque to check against be loaded internally by using the  sp.loadLastReceivedCheque() function instead of being passed as a parameter?

Reply:
We can make it a package level function. I wouldn't load the last received cheque internally because then it makes testing this function in isolation harder.

Reply:
this was fixed